### PR TITLE
Adding bootstrap sync strategy for empty lease table

### PR DESCRIPTION
**Pull request by jushkem**
Tested with 200KB record size with SubscribeToShard, 10 seconds process time, 10 shards and monitored durability metrics.

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java [hunk 1]
```diff
@@ -253,8 +255,32 @@ private synchronized void assertHashRangeOfClosedShardIsCovered(final Shard clos
...
+            case AT_TIMESTAMP:
+                builder = builder.type(ShardFilterType.FROM_TIMESTAMP).timestamp(initialPositionInStreamExtended.getTimestamp().toInstant());
+                break;
+        }
+
+        return builder.build();
+
+    }
+
+    private static List<Shard> getShardList(@NonNull final ShardDetector shardDetector, InitialPositionInStreamExtended initialPositionInStreamExtended)
```
Comment (ashwing) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java]: We need both getShardList(detector) and getShardList(detector, initialPosition). Former for non-empty lease table case and latter for empty lease table

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java [hunk 2]
```diff
@@ -78,7 +80,7 @@ public synchronized void checkAndCreateLeaseForNewShards(@NonNull final ShardDet
             final boolean cleanupLeasesOfCompletedShards, final boolean ignoreUnexpectedChildShards,
             final MetricsScope scope) throws DependencyException, InvalidStateException,
             ProvisionedThroughputException, KinesisClientLibIOException {
-        final List<Shard> latestShards = getShardList(shardDetector);
+        final List<Shard> latestShards = getShardList(shardDetector, initialPosition);
```
Comment (ashwing) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java]: We need additional information from outside this class, to use shardfilter or not. We will be using shardfilter only in empty lease table case.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/KinesisShardDetector.java [hunk 3]
```diff
@@ -172,13 +179,13 @@ public Shard shard(@NonNull final String shardId) {
...
 
-    private ListShardsResponse listShards(final String nextToken) {
+    private ListShardsResponse listShards(ShardFilter shardFilter, final String nextToken) {
         final AWSExceptionManager exceptionManager = new AWSExceptionManager();
         exceptionManager.add(LimitExceededException.class, t -> t);
         exceptionManager.add(ResourceInUseException.class, t -> t);
         exceptionManager.add(KinesisException.class, t -> t);
 
-        ListShardsRequest.Builder request = KinesisRequestsBuilder.listShardsRequestBuilder();
+        ListShardsRequest.Builder request = KinesisRequestsBuilder.listShardsRequestBuilder().shardFilter(shardFilter);
```
Comment (ashwing) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/KinesisShardDetector.java]: Am assuming setting null shardFilter is valid. Did you verify this from any coral doc or testing?

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java (hunk 2 above)
Comment (jushkem) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java]: Okay. Will add information from leaseRefresher to verify if the leaseTable is empty or not.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java (hunk 1 above)
Comment (jushkem) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java]: Yeah, will add info from LeaseRefresher.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/KinesisShardDetector.java (hunk 3 above)
Comment (jushkem) [reply amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/KinesisShardDetector.java]: Yeah `null` ShardFilter is equivalent to making an ordinary ListStreams call.

File: amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java [hunk 4]
```diff
@@ -253,8 +255,34 @@ private synchronized void assertHashRangeOfClosedShardIsCovered(final Shard clos
...
+
+        switch (initialPositionInStreamExtended.getInitialPositionInStream()) {
+            case LATEST:
+                builder = builder.type(ShardFilterType.AT_LATEST);
+                break;
+            case TRIM_HORIZON:
+                builder = builder.type(ShardFilterType.FROM_TRIM_HORIZON);
+                break;
+            case AT_TIMESTAMP:
+                builder = builder.type(ShardFilterType.FROM_TIMESTAMP).timestamp(initialPositionInStreamExtended.getTimestamp().toInstant());
```
Comment (jushkem) [inline amazon-kinesis-client/src/main/java/software/amazon/kinesis/leases/HierarchicalShardSyncer.java]: Synced offline - need to change this to `AT_TRIM_HORIZON` and `AT_TIMESTAMP`, we should avoid client-side processing where possible and use child shard information to discover newer shards.

**Comment by jushkem** [comment]
Moved this PR to ashwin's private branch.

**Pull request closed without merging**